rfcs/convergence/internationalization.md

```diff
@@ -0,0 +1,168 @@
+
+```js
+var localeValues = {
+  locale: 'es',
+  global: {
+    placeholder: 'Seleccione'
+  },
+  Table: {
+    filterTitle: 'Filtrar menú',
+    emptyText: 'Sin datos',
+    selectAll: 'Seleccionar todo',
+    [...]
+  },
+  Modal: {
+    okText: 'Aceptar',
+    cancelText: 'Cancelar',
+    justOkText: 'Aceptar'
+  }
+  [...etc]
+}
+```
+
+Material UI includes the locale strings in their ThemeProvider, and Ant Design puts it in their ConfigProvider.
+
+Libraries where I found no tools or defined patterns for localization include Lightning Design System, Semantic UI, React Bootstrap, and Carbon Design System.
+
+Two of the most common i18n packages for React independent of component libraries are `react-i18next` (a react wrapper for i18next), and `react-intl`. Both provide a helper function to format a string -- at the most basic, they take a key and return a localized string. The most basic usage looks like this for each library:
+
+react-i18next:
+
+```js
+import { useTranslation } from 'react-i18next';
+
+const Component = () => {
+  const { t } = useTranslation();
+
+  return <div>{t('key')}</div>;
+};
+```
+
+react-intl:
+
+```js
+import { useIntl } from 'react-intl';
+
+const Component = () => {
+  const intl = useIntl();
+
+  return <div>{intl.formatMessage({ id: 'key' })}</div>;
+};
+```
+
+Both also provide extras like React components as an alternative to the function, and helpers for things like dates, string template value replacement, SSR, etc. Those likely aren't relevant to Fluent, since we'll either not use them in the case of the component or SSR, or provide our own solution for dates/string templates.
+
+## Proposal
+
+While we don't have plans to provide actual translations for the strings in any of our components, it'd be nice to create a pattern for defining custom strings that makes it easy for authors to:
+
+- Easily propagate app-level locale strings down to Fluent components without needing to manually map them to multiple separate component props
+- Easily include all necessary component strings when they create a new locale definition/JSON file (i.e. not need to manually remember that TagPicker needs `selectionAriaLabel` defined).
+- Use 3rd party i18n tools with Fluent components without too much extra work
+
+To do this, I think it makes sense to add locale data to Fluent's `ProviderContext`, since it already includes `dir`.
+
+One straightforward option for how to do this would be to provide a locale object with static key/value pairs for each string. We could then import defaults for all strings where defaults make sense, which would allow us to also share global strings for things like 'close'/'remove'/'select all'/etc. So something like this (specific file names and values are just for show):
```

Comment:
I don't think we should have a single file for all strings:
1. Bad for bundle size. The extreme example is once we converge Calendar/DatePicker, that has a LOT of strings, but relatively few pages will need that component (especially on initial page load).
2. It's against the patterns we've used so far to have one low-level package which also knows about individual components.

But standardizing a location/name for default strings within each package would make sense. I also agree it might make sense to have very common strings defined once in a common location.

Reply:
Agree there with @ecraig12345. We have an example of such objects in Northstar:

![image](https://user-images.githubusercontent.com/14183168/129880198-ff827199-4a74-4cdd-8468-76db6af784a4.png)
_**GZIP** sizes of individual exports_


- Themes in that case are not tree-shakable i.e. include styles even for unused components
- Without a valid theme components will not have styles i.e. increases initial page load

For a locale object it will be the same thing as every additional string that will be added - will increase bundle size for every component 😥

---

For example, `react-spectrum` handles such strings per component: https://github.com/adobe/react-spectrum/tree/main/packages/%40react-spectrum/pagination.

In the same, we should keep config simple for consumers. `FluentProvider` can host a name of locale and use similar locale detection as in: https://github.com/adobe/react-spectrum/blob/1e2b7f280a04ead83e21b27fa580eb189e5f4186/packages/%40react-aria/i18n/src/useDefaultLocale.ts#L47-L50

_BTW, SSR scenario is handled there._

Reply:
Ah yeah, Calendar/Datepicker's a great example of something we wouldn't want to ship all the time. Aside from date-time, I actually think we won't have many instances of default values for strings that are component-specific.

Assuming that's the case, how would you feel about one single file for global defaults (e.g. "close", "open", "filter", "{0} results", etc.)? We could start with that, and see if we even need to have component-specific default strings. If we do, we could start by importing those defaults directly from an in-package string default file in addition to the global default file. Since that would be an internal implementation detail, I don't think we'd need to worry too much about getting it perfect to start. As long as an author can always pass in strings through a `strings` prop, how we organize our defaults shouldn't be a breaking change.

Does that sound like a good approach to you, and would that approach cover your bundle size and pattern concerns?

Reply:
> Aside from date-time, I actually think we won't have many instances of default values for strings that are component-specific.
 
ColorPicker 😆 but that's not scheduled any time soon

Reply:
> Assuming that's the case, how would you feel about one single file for global defaults (e.g. "close", "open", "filter", "{0} results", etc.)?

If you plan to add these defaults to `FluentProvider`, this will still increase bundle size for component like `Button` or `Text` that don't need these defaults.

Reply:
There's a bit of a tradeoff since it would slightly increase the bundle size for someone using only the Button or Text components, but would decrease the size for someone using multiple components like Dialog, Combobox, SplitButton, List, etc. I'm also guessing the entire global default strings file won't be more than ~20 lines at a generous estimate, based on the Material and Ant Design precedents, so not a huge bundle size impact either way :).

The other benefit of having a set of global defaults is that it will ensure that there is only one string name for something like "close". That way a team translating our components won't need to worry about mapping the same translated string to both `strings.close` and `strings.closeButtonLabel`, just because we didn't manually check that our keys are consistent across all components.